### ISSUE TITLE
Optimize core calls to `file_directory_temp()`.

### ISSUE THREAD

**Issue by alanmels** [backdrop/backdrop-issues#4199]
**Description of the bug**

Per discussion on https://github.com/backdrop/backdrop-issues/issues/4193#issuecomment-550064054 there are two different PRs for the same issue:

1. The one on https://github.com/backdrop/backdrop/pull/2972 is to save found temporary directory within the `file_directory_temp()` function itself;

2. The one on https://github.com/backdrop/backdrop/pull/2977 is to do the same in the `system.install`, because @indigoxela is sure it must be fixed there.

Now, if any of those two approaches go in and the https://github.com/backdrop/backdrop-issues/issues/4193 is resolved, **then and only then** the following idea gets its actuality:

> @jenlampton
> > When is `file_directory_temp()` called? I'm guessing when the form is saved, and on install? Any other time?
> 
> @alanmels
> > According to https://api.backdropcms.org/api/backdrop/core%21includes%21file.inc/function/calls/file_directory_temp/1 there are six calls to the function:
> > 
> > 1. BackdropTemporaryStreamWrapper::getDirectoryPath in core/includes/stream_wrappers.inc
> >    Implements abstract public function getDirectoryPath()
> > 2. ConfigurationUITest::testExport in core/modules/config/tests/config.test
> >    Tests export of configuration.
> > 3. config_download_full_export in core/modules/config/config.admin.inc
> >    Downloads a tarball of the site configuration.
> > 4. FileDirectoryTest::testFileDirectoryTemp in core/modules/simpletest/tests/file.test
> >    Ensure that the file_directory_temp() function always returns a value.
> > 5. system_file_system_settings in core/modules/system/system.admin.inc
> >    Form builder; Configure the site file handling.
> > 6. system_requirements in core/modules/system/system.install
> >    Implements hook_requirements().
> > 
> > So the last one shows it gets called during the installation. That's why on a system that was just installed you see the path on UI, but the config file is not aware of the path.
> 
> @jenlampton 
> > It looks like a few of these might need to be updated to use `config_get()`. I'm thinking `config_download_full_export()` and `ConfigurationUITest::testExport` for sure. Maybe also `FileDirectoryTest::testFileDirectoryTemp`?
> 
> @alanmels
> >I agree: once we have the temporary directory set in the configuration file, there is no need to call `file_directory_temp()` at all. I could take a stab for another PR to incorporate all suggestions, but shouldn't we address this single issue first? Taking into account it is a bug, indeed!

That's why I am creating this issue separately from https://github.com/backdrop/backdrop-issues/issues/4193, because that one is to provide solution for the existing discrepancy between UI (that shows temporary directory path) and the configuration file (that shows empty temporary file), while this one is to replace all the unnecessary occurrences of `file_directory_temp()`.

**Comment by alanmels** [comment]
I have not modified only item 4 in the list:

> FileDirectoryTest::testFileDirectoryTemp in core/modules/simpletest/tests/file.test
> Ensure that the file_directory_temp() function always returns a value.
> 

because that function seems to only simulate different scenarios for testing purposes. 

For the rest I've replaced the occurrences of `file_directory_temp()` with `config_get('system.core', 'file_temporary_path')`, because IMO ideally the directory should be set once during the installation and then, even if it's changed via UI later, should be retrieved from the configuration file. So no need to call `file_directory_temp()` from anywhere else.

Correct me if I am wrong, because I must admit I had no time to track the logic behind of each use of `file_directory_temp()` in the above functions.

**Comment by herbdool** [comment]
I don't think this is necessary and it may even cause more issues since there is no more fallback if the variable is removed.

In https://www.drupal.org/project/drupal/issues/930122#comment-3524752 the dev says:

> Coder Review module recommends to use the function file_directory_temp() directly, and i don't see any reason not to follow that advice throughout core. In system.install it implies that '$file_temporary_path' could be set in settings.php distinct from '$file_directory_temp', but there is no documentation of such a concept in default.settings.php or anywhere else, and this code would be equally functional (in refraining from triggering the file_directory_temp() function) in using the $file_directory_temp variable.
>
>Also, a reference to checking file_temporary_path on line 2093 of system.module should be removed– it doesn't check it nor does it check file_directory_temp, and it's third in a list that begins "both ..." anyway ;-)
>
>In short, i am in no way a file system expert, but i think all core code should use the function file_directory_temp() everywhere, except system.install, which wants to avoid calling that function during the install phase, and so should use the 'file_directory_temp' variable directly.
>
>Please banish 'file_temporary_path'!
